### DISCUSSION TITLE
What is the best practice to exclude certain targets when in test?

### DISCUSSION THREAD

**Discussion by anddrzejb** [dotnet/msbuild#6619] [Q&A] [answered]
I have several blazor projects that reference the same c# library. That library runs some gulp tasks, that take a long time (compilation of css, sass, etc). These are completely useless during tests, so I would like to skip them, when tests are executed.

Now I tried with passing a variable to the project like this from my `Tests.csproj`
```xml
<Project Sdk="Microsoft.NET.Sdk.Razor" InitialTargets="InformAboutTest">

  <Target Name="InformAboutTest" BeforeTargets="Build">
    <MSBuild Projects="MyProject.csproj" Properties="IsUnderTest=true;" />
  </Target>
```
I added this to `MyProject.csproj`:
```xml
  <Target Name="DebugRunGulp" BeforeTargets="DebugEnsureNodeEnv" Condition=" '$(Configuration)' == 'Debug' And Exists('$(SolutionDir)node_modules') ">
    <Message Importance="high" Text="MyProject, IsUnderTest=$(IsUnderTest)"/>
    <Exec WorkingDirectory="$(SolutionDir)" Command="npm run gulp:components" />
  </Target>
```
and when I run the tests I get this output:
```log
❯ dotnet test
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  MyProject-> D:\dotnet\MyProject\bin\Debug\net6.0\MyProject.dll
  MyProject, IsUnderTest=true
  Determining projects to restore...
C:\Program Files\dotnet\sdk\6.0.100-preview.4.21255.9\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error NETSDK1005: Assets file 'D:\dotnet\MyProject\obj\project.assets.json' doesn't have a target for 'net6'. Ensure that restore has run and that you have included 'net6' in the TargetFrameworks for your project. [D:\dotnet\MyProject\MyProject.csproj]
```
It seems that `MyProject.csproj` picks up the property, and I think I can use it, but whatever I did until now, I cannot get the project to actually successfully build. If I remove either the whole target `InformAboutTest` or remove it from `InitialTargets`, then everything builds and tests run. I also tried with adding to my target `BeforeTargets="PreBuildEvent"` but that does not work (`MyProject` is built before `Tests` and never gets the property). What am I missing?

**Comment by rainersigwald** [comment, marked as answer]
A few comments:

> These are completely useless during tests, so I would like to skip them, when tests are executed.

Can you instead add incremental build support for those things so that they are skipped when not necessary, either with MSBuild-level `Inputs` and `Outputs` around the `Task` or with Gulp's incremental functionality? That will help for non-test build scenarios too.

> I tried with passing a variable to the project like this from my `Tests.csproj`
> 
> ```
> <Project Sdk="Microsoft.NET.Sdk.Razor" InitialTargets="InformAboutTest">
> 
>   <Target Name="InformAboutTest" BeforeTargets="Build">
>     <MSBuild Projects="MyProject.csproj" Properties="IsUnderTest=true;" />
>   </Target>
> ```

Does this project also have a `ProjectReference` to `MyProject.csproj`? If so, you've created a _second_ reference to the project, which isn't what you'd want, and you're getting the failure because you're not getting the full power of `ResolveProjectReferences` (specifically here it looks like your two projects have different TargetFrameworks, so the reference needs to specify that explicitly).

![image](https://user-images.githubusercontent.com/3347530/123142212-ca1df380-d41e-11eb-98bb-2ab211a16fd0.png)

----

> ⚠ Don't do this, explanation below.

The most direct way to implement what you're doing is to tell the ProjectReference to pass your special property along:

```xml
<ProjectReference Include="..\MyProject\MyProject.csproj" AdditionalProperties="IsUnderTest=true" />
```

HOWEVER, that will cause a build race condition if you ever build a solution or traversal project that references both your test project and your library, because the solution->library reference will race with the "special" test->library reference.

**Reply by anddrzejb** [reply]
Thank you for your help! I will try to do the incremental build solution. However, my both projects target **only** net6. The message I was getting was very confusing and misleading I think...
